File: Structural/Proxy/Delphi/RealSubjectImp.pas
unit RealSubjectImp;

interface

uses
  RealSubject;

type
  TRealSubject = class(TInterfacedObject, IRealSubject)
  public
    constructor Create();
    destructor Destroy(); override;
    function GetData(): string;
  end;

implementation

constructor TRealSubject.Create();
begin
  inherited Create();
  Writeln('Connection created');
end;

destructor TRealSubject.Destroy();
begin
  Writeln('Connection is closed.');
  inherited Destroy();
end;

function TRealSubject.GetData(): string;
begin
  Result := 'Data received';
end;

end.

File: Structural/Proxy/C#/ProxyExample/ProxyExample/RealSubjectImp.cs
namespace ProxyExample;

internal class RealSubject : IRealSubject
{
	public RealSubject()
	{
		Console.WriteLine("Resource created");
	}

	~RealSubject()
	{
		Console.WriteLine("Resource released");
	}

	public string GetData()
	{
		return "Data received";
	}
}